### ISSUE TITLE
liveCD user/password issue ubuntu mini remix 12.04

### ISSUE THREAD

**Issue by hreikin** [kamilion/customizer#45]
when the iso is made and burned to a disk during installation when using the "Try OS" option in ubiquity you get to the login screen enter username specified when using customizer with no password and select session then hit enter. it returns incorrect password error ?

the username was set to "liveuser" so i dont think it could be this issue affecting it : 

https://docs.google.com/document/d/1PfhHnSBjv-IDI7Yh5obhMGYCAV9Gw1NPEynU4GqKTsA/edit#bookmark=id.4kp1lixr2n23

ive tried different usernames such as the hostname i set and the name i chose for my liveCD (just on the off-chance) but none work, this has happened on a few liveCDs ive made but not all and as far as i can tell i havent done anything different (apart from choosing different names but never any with anything other than letters in) i can still install from ubiquity i just cant try the live system beforehand as i cant log in

my desktop is running a fresh install of 32 bit ubuntu 12.04 from here : http://releases.ubuntu.com/precise/
im using the 32 bit ubuntu 12.04 mini remix in customizer from here : http://www.ubuntu-mini-remix.org/
customizer version : 3.2.3

i installed customizer with this method : 

wget https://dl.dropboxusercontent.com/u/54183088/install.sh
chmod +x install.sh
sudo ./install.sh -i

any help would be appreciated

EDIT : im not sure if customizer is mixing up the live username and the hostname or OS name as i made another liveCD last night with all 3 set to the same name ("our" in this case for my openbox ubuntu remix) and the liveCD username worked, ill try make another later today and see if i can figure out if it is mixing the names up.

**Comment by hreikin** [comment]
right here were the settings on screen for my OUR iso i made : 

distribution name = OUR
liveCD user = our
hostname = our

it let me log in to my liveCD with the username our, so i opened customizer and changed those 3 things only to this : 

distribution name = aaa
liveCD user = bbb
hostname = ccc

now when i tried to log in to the liveCD of this version i had to use the distribution name, aaa, which i found odd as when i logged into the OUR version the distribution name was in caps but the user login for the liveCD was lowercase

i have no idea why it is doing this and its quite confusing, below is a screenshot of my customizer as i set it up for my test

![customizer screenshot](https://f.cloud.github.com/assets/6406513/1934656/fdcd610e-7eed-11e3-842f-8f2bd2bb2513.png)

any ideas ?

**Comment by ghost** [comment]
@hreikin 
Hi, there! I have been trying out using Ubuntu Mini Remix 12.04 to reproduce your issue.
And I can confirm your issue is valid.

In my case, I only change the LiveCD username and hostname from Customizer 3.2.3 GUI. I rebuild ISO with minimal XFCE4 packages with LightDM login manager and other essential packages. The LiveCD only shows up the login manager with hostname "ubuntu", which is the default. Typing "ubuntu" in username and press `Enter`, then without password (and press `Enter` again) will bring me into Live Desktop successfully.

This issue will be investigated.

@fluxer 
I believe this is _not_ related to previous issue #25 which has been solved earlier.
What I can say from my side:
- Customizer GUI did write changes to `/FileSystem/etc/casper.conf` file (shown below are changes I have made to USERNAME and HOST)
  
  ```
  export USERNAME="remix"
  export USERFULLNAME="Live session user"
  export HOST="live"
  export BUILD_SYSTEM="Ubuntu"
  ```
- But the rebuilded ISO had its `/etc/casper.conf` file remain unchanged ( shown below are existing casper.conf that has been confirmed from XFCE4 Terminal on LiveCD )
  
  ```
  export USERNAME="ubuntu"
  export USERFULLNAME="Live session user"
  export HOST="ubuntu"
  export BUILD_SYSTEM="Ubuntu"
  ```

This may mean, the modified casper.conf file has failed to copy to ISO during building, or somehow wasn't included in rebuilding process.

Could you help us look into this issue as well?

**Comment by fluxer** [comment]
@clearkimura Probably. I will need the log from the terminal when you run Customizer. Just run it and
change the hostname, distro name, etc once so that the events that should replace the variables in
FileSystem/casper.conf are triggered. It will tell me what's going when that happens with many details.
I knew someday this issues will happen so I added lots of debug messages :).

**Comment by ghost** [comment]
@fluxer 
Open Terminal, run Customizer GUI, select previously built ISO, edit LiveCD username and hostname from "remix" and "live" then change to "test" and "again", finally rebuild. Pasted log here: http://pastebin.com/QE3uF48h

**Comment by hreikin** [comment]
@clearkimura thanks for the response, if there is anything i can do to help then just let me know

**Comment by fluxer** [comment]
@clearkimura @hreikin, can any of you cat > me the whole casper.conf file? I think there was something about the 12.x releases where a line should be uncommented for the changes to apply, otherwise the defaults are used.

**Comment by fluxer** [comment]
Or was it in etc/lsb-release, cat > me that one too if you can.

**Comment by ghost** [comment]
@fluxer 
The LiveCD system is minimal, I can't copy-paste or transfer content from virtual machine to host by any means. So I pasted screenshot as below:

![20140118-024727-scrot](https://f.cloud.github.com/assets/1223304/1943283/1346880c-7fa8-11e3-8192-a6fb8064b451.png)

And copy of output from ISO before rebuild, via Customizer Terminal:

```
(chroot)root@xxxxx:/# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=12.04
DISTRIB_CODENAME=precise
DISTRIB_DESCRIPTION="Ubuntu 12.04"

(chroot)root@xxxxx:/# cat /etc/casper.conf
# This file should go in /etc/casper.conf
# Supported variables are:
# USERNAME, USERFULLNAME, HOST, BUILD_SYSTEM, FLAVOUR

export USERNAME="test"
export USERFULLNAME="Live session user"
export HOST="again"
export BUILD_SYSTEM="Ubuntu"

# USERNAME and HOSTNAME as specified above won't be honoured and will be set to
# flavour string acquired at boot time, unless you set FLAVOUR to any
# non-empty string.

# export FLAVOUR="Ubuntu"

(chroot)root@xxxxx:/# 
```

Will these do?

**Comment by fluxer** [comment]
@clearkimura Thanks,. FLAVOUR should be exported but before doing anything else can you @hreikin confirm what's the situation with it on Linux Mint?

```
grep FLAVOUR /home/FileSystem/etc/casper.conf
```

**Comment by hreikin** [comment]
@fluxer i'll download linux mint now and give it a whirl, and then post pastebins/screenshots of relevant info

**Comment by hreikin** [comment]
@fluxer i hope this helps :)

i cleaned my directory then closed customizer and started it again, using this command :

gksu /opt/Customizer/GUI.gambas

i then selected the linux mint 16 cinnamon ISO then ran the 2 cat commands and changed the liveuser to aaa and the hostname to bbb and then clicked terminal and ran these 2 cat commands again (pastebins attached for the commands ran before and after i changed the liveuser/hostname) : 

cat /etc/lsb-release 
cat /etc/casper.conf 

terminal output before i changed the details = http://pastebin.com/dG4qKWxK
terminal output after i changed the details = http://pastebin.com/nkAXMM0K

then i rebuilt the ISO and ran it in qemu (which failed) so i installed virtualbox and tried to load it in there, here is the screenshot of those 2 commands ran in the live environment (please note it automatically logged in as aaa and gave a notice as well) : 

![linuxmintcustomizerscreenshot](https://f.cloud.github.com/assets/6406513/1945161/06c577f8-7fc8-11e3-8a2f-50aef8784a23.png)

i then quit the emulator and ran this command in a terminal in customizer : 

grep FLAVOUR /home/FileSystem/etc/casper.conf

then in a terminal in customizer ran the 2 cat commands again, here is the pastebin (including the grep command, did i do it in the wrong place cos it gave an error ?) : http://pastebin.com/ZZq3n86U

and then rebuilt the ISO once more and ran it in the live environment again, here is the screenshot of the cat commands ran in virtualbox after i rebuilt the ISO for the 2nd time : 

![linuxmintcustomizerscreenshot1](https://f.cloud.github.com/assets/6406513/1945368/48fe5690-7fcc-11e3-8347-3ca1532d1134.png)

and finally this is the pastebin from the terminal which was running customizer the whole time : http://pastebin.com/0Y8gc35M

i hope i did everything in the right order, if not let me know and ill run it again

**Comment by fluxer** [comment]
@hreikin Much appreciated! I will work on the issue soon, I know what I have to do - uncomment "export FLAVOUR" in casper.conf, other than that Customizer works as expected. Linux Mint is not affected by this as the devs have done it for you so their changes apply so this is a quirk only for official (or close to) Ubuntu ISOs.

Thanks again!

**Comment by hreikin** [comment]
@fluxer ok cool, i look forward to the update

**Comment by fluxer** [comment]
Should be fix in both master and gambas3 branches. I expect issues with the Linux Mint installer specificly since FLAVOUR will not be "mint" but who knows. Can you test if it works as it should?

**Comment by hreikin** [comment]
ill try on ubuntu and then on mint and get back to you, do you want me to run the cat commands again before and after i change the liveuser/hostname and when i emulate it ?

**Comment by fluxer** [comment]
Well, why not. Just to make sure.

**Comment by hreikin** [comment]
ok ill do this now and then post the pastebins/screenshots

**Comment by fluxer** [comment]
@hreikin Did you install the latest version of Customizer? Run the install script.
I do not see the FLAVOUR being appended in casper.conf.

The size of the ISO probably changed because Customizer uses XZ compression
for the root filesystem where the official images do not I guess.

**Comment by hreikin** [comment]
yeh, i had that thought about 10 mins after i left home (i blame lack of sleep !) gimme a couple of hours to do it all and i will repost everything, sorry for the screw up :D

ill delete the other post once ive made the new one

**Comment by hreikin** [comment]
@fluxer right this one should be correct :D for this i used official ubuntu 12.04 and linux mint 16 cinnamon (both 32-bit),i started with a clean install (i messed up my HD earlier and had to reinstall my host system), i loaded an ISO then opened the terminal in customizer, these are pastebins of cat commands for both OS's before i changed anything : 

ubuntu = http://pastebin.com/Jf5Hu5hP
mint = http://pastebin.com/HbCkn9LW

pastebins of cat commands after i changed liveuser and hostname (changed to aaa and bbb in both OS's) :

ubuntu = http://pastebin.com/XtpK7xwM
mint = http://pastebin.com/pWcPvFik

i then built the ISO and emulated them (used the"Try OS" option rather than installing to VB), screenshots of cat commands ran in virtualbox for both OS's :

ubuntu =

![ubuntuvbcatcommands](https://f.cloud.github.com/assets/6406513/1948126/cedb47a8-807d-11e3-92b9-c197e15b971b.png)

mint =

![mintcustomizervbcatcommands](https://f.cloud.github.com/assets/6406513/1948192/b1fdf87e-8082-11e3-858b-452b83dbb98a.png)

another note, in the terminal in ubuntu in customizer/when emulated the (chroot)root@ubuntu-32:/# line after i ran the cat /etc/casper.conf command wasn't on a new line like it was previously. you should see in the screenshot.

**Comment by ghost** [comment]
@fluxer 
I agree with @hreikin as I also experience "without the newline" in Terminal running on Ubuntu. Below is my screenshot of output from Ubuntu Mini Remix 12.04 with minimal XFCE system (the same ISO, whose username and hostname renamed to "remix" and "live", then rebuild using latest updated Customizer).

![20140119-132242-scrot](https://f.cloud.github.com/assets/1223304/1948935/b80b3640-80c9-11e3-96da-6057906a7fe1.png)

On side note, my earlier explanation of "unable to auto-login into LiveCD" is due to my default LightDM configuration files. The auto-login issue has nothing to do with @hreikin username/hostname issue.

**Comment by fluxer** [comment]
I will fix the new-line problem right away. Can we close this now?

**Comment by ghost** [comment]
@fluxer 
I have tested Customizer from both master and gambas3 branch, one by one.

Either case, your attempt fix as `export FLAVOUR="custom\n"` doesn't work, because the result of fix includes `\n` within a string. So in the Live session, the bash will display exactly same with additional `\n` in output, instead of a newline.

Seemed like the newline isn't defined at the first place.

I did a quick lookup on gambasdoc.org, and [found this page](http://gambasdoc.org/help/comp/gb/gb/newline). If I understand correctly, the newline definition should appear in the target `.class` file.

Note the minor difference between gambas version 2 and 3:

[Gambas 2 NewLine](http://gambasdoc.org/help/comp/gb/gb/newline?v2)

```
 CONST NewLine AS String = "\n"
```

[Gambas 3 NewLine](http://gambasdoc.org/help/comp/gb/gb/newline?v3)

```
 Const NewLine As String = "\n"
```

In short, I think we should include the newline definition in `FMain.class` file.

**Comment by fluxer** [comment]
I do not have the tools to test this, but you do ;). If you can push the changes that would be great.

**Comment by ghost** [comment]
@fluxer 
My apologies, I don't have time to test now (this weekend) as I have to leave home earlier to attend co-worker guests. Anyway, as termporal measure, I will write a simple workaround for users as below.

@hreikin 
Until I can test and push the fix, I can suggest a simple workaround for you.

Before you start rebuilding the ISO, do the following:
1.  In Customizer, click on `Directories` > `FileSystem`. It will open a File Manager as root at target path.
2.  Double-click on `etc` folder
3.  Look for `casper.conf` file and open in a Text Editor.
4.  Go to the last line where you can find `export FLAVOUR="custom\n"`
5.  Delete the `\n`
6.  Then, put your cusor at the end of this line and press `Enter` key to insert a newline.
7.  Finally, save the file

Then, rebuild ISO and run in a virtual machine. After Live session Desktop is loaded, run Terminal and type `cat /etc/casper.conf` and press `Enter`. You will see the output with newline at the end, as shown below.

![20140125-111819-scrot](https://f.cloud.github.com/assets/1223304/2001366/cc730d1c-85a4-11e3-9541-b76d1d4c81d8.png)

You will need to do this only once per loaded ISO image.

For time being, please spare me with this workaround.

**Comment by hreikin** [comment]
@clearkimura thats great, not a problem, started a 2nd job last week so i know how you feel with time limitations :) thanks for the workaround :)

**Comment by ghost** [comment]
@fluxer 
I tried to fix this newline issue, using local source codes on my PC. And I failed to make any changes in Gambas2 IDE. The menu--Compile, Compile All, Update all forms--didn't seemed to bring any changes.

Since I have no idea about Gambas programming, I suppose I don't know how to update the code properly.  Sorry, I am useless in this case. This issue will remain open.

@hreikin 
That said, please spare me with the workaround I explained to you earlier.

**Comment by colmragu** [comment]
Hay I have a similar problem to this. 
I would edit casper.conf then the rebuild casper scipts would fail like 

[_] Making sure everything is configured
Reading package lists...
Building dependency tree...
Reading state information...
0 upgraded, 0 newly installed, 0 to remove and 208 not upgraded.
update-initramfs: Generating /boot/initrd.img-3.11.0-17-generic
cryptsetup: WARNING: could not determine root device from /etc/fstab
/tmp/mkinitramfs_HvV1lh/scripts/casper-bottom/48xubuntu_maybe_ubiquity: 6: .: Can't open /scripts/casper-functions
[_] Unblocking files

I solved it by running 
"ln -s /usr/share/initramfs-tools/scripts"
inside the chroot environment.

Should I add another bug or leave it with this one. I can push a patch for it if this solves for others.

**Comment by fluxer** [comment]
@colmragu Your problem has nothing to do with this issue, patches welcome tough. Just fork and make pull request.

Cheers!

**Comment by ghost** [comment]
This is another old issue that I didn't notice until today. As per commit f9a0d52, this issue has been fixed already in master branch (in late 2014).

Except for gambas3 branch, there is a leftover issue that produce an invalid newline in `casper.conf` file. But since gambas3 branch is unsupported, this issue shall be spared with the given workaround as noted in [this comment dated Jan 25, 2014](https://github.com/clearkimura/Customizer/issues/45#issuecomment-33285205) as above.

To be sure that master branch indeed not having this issue, I have tested two ISO images i.e. Ubuntu Mini Remix 14.04 and Xubuntu 14.04 using the latest release of Customizer 4.1.1. I opened chroot Terminal, run `cat /etc/casper.conf` and confirmed that the output includes a newline.

The correct behaviour of output can be seen in the screenshot found in the linked comment. Following this, since master branch is no longer affected by this issue, this issue will be closed.
